contrib/endpoints/src/api_manager/firebase_rules/firebase_request.cc

```diff
@@ -0,0 +1,402 @@
+      next_request_(nullptr) {
+  firebase_http_request_.url =
+      firebase_server_ + kV1 + "/" + ruleset_name + kTestQuery;
+  firebase_http_request_.method = kHttpPostMethod;
+  firebase_http_request_.token_type =
+      auth::ServiceAccountToken::JWT_TOKEN_FOR_FIREBASE;
+  external_http_request_.token_type =
+      auth::ServiceAccountToken::JWT_TOKEN_FOR_AUTHORIZATION_SERVICE;
+
+  // Update the first request to be sent which is the TestRulesetRequest
+  // request.
+  SetStatus(UpdateRulesetRequestBody(RepeatedPtrField<FunctionCall>()));
+  if (!current_status_.ok()) {
+    return;
+  }
+
+  next_request_ = &firebase_http_request_;
+}
+
+bool FirebaseRequest::is_done() { return is_done_; }
+
+HttpRequest FirebaseRequest::GetHttpRequest() {
+  if (is_done()) {
+    return HttpRequest();
+  }
+
+  if (next_request_ == nullptr) {
+    SetStatus(Status(Code::INTERNAL, "Internal state in error"));
+    return HttpRequest();
+  }
+
+  return *next_request_;
+}
+
+Status FirebaseRequest::RequestStatus() { return current_status_; }
+
+void FirebaseRequest::UpdateResponse(const std::string &body) {
+  if (is_done()) {
+    env_->LogError(
+        "Receive a response body when no HTTP request is outstanding");
+    return;
+  }
+
+  if (next_request_ == nullptr) {
+    env_->LogError(
+        "Received a response when there is no request set"
+        "and when is_done is false."
+        " Looks like a code bug...");
+    SetStatus(Status(Code::INTERNAL,
+                     "Internal state error while processing Http request"));
+    return;
+  }
+
+  Status status = Status::OK;
+
+  // If the previous request was firebase request, then process its response.
+  // Otherwise, it is the response for external HTTP request.
+  if (next_request_ == &firebase_http_request_) {
+    status = ProcessTestRulesetResponse(body);
+  } else {
+    status = ProcessFunctionCallResponse(body);
+  }
+
+  if (status.ok()) {
+    status = SetNextRequest();
+  }
+
+  SetStatus(status);
+  return;
+}
+
+void FirebaseRequest::SetStatus(const Status &status) {
+  if (!status.ok() && !is_done_) {
+    current_status_ = status;
+    is_done_ = true;
+  }
+}
+
+// Create the TestRulesetRequest body.
+Status FirebaseRequest::UpdateRulesetRequestBody(
+    const RepeatedPtrField<FunctionCall> &function_calls) {
+  proto::TestRulesetRequest request;
+  auto test_case = request.mutable_test_suite()->add_test_cases();
+  test_case->set_expectation(proto::TestCase::ALLOW);
+
+  ::google::protobuf::Value token;
+  ::google::protobuf::Value claims;
+  ::google::protobuf::Value path;
+  ::google::protobuf::Value method;
+
+  Status status = utils::JsonToProto(context_->auth_claims(), &claims);
+  if (!status.ok()) {
+    return status;
+  }
+
+  auto *variables = test_case->mutable_request()->mutable_struct_value();
+  auto *fields = variables->mutable_fields();
+
+  path.set_string_value(context_->request()->GetRequestPath());
+  (*fields)[kPath] = path;
```

Comment:
No. SetProtoValue will not work here. This is because of the the way mutable_* works. Since I need all the kPath and kAuth in the same map, I cannot use SetProtoValue.